File: rust/bifti/src/loader.rs
use std::{
    collections::HashMap,
    path::{Path, PathBuf},
    rc::Rc,
};

use nifti::{NiftiObject, NiftiVolume, ReaderStreamedOptions};

use crate::{
    BiftiPhantom, BiftiTissue, NiftiRef, ResliceTo,
    phantom::TissueProperty,
    resample::Resampler,
    volume::{Volume, VolumeData, map_variant},
};

pub struct Phantom {
    pub config: BiftiPhantom,
    pub tissues: HashMap<String, Tissue>,
}

impl Phantom {
    #[cfg_attr(
        feature = "tracing",
        tracing::instrument(skip_all, fields(path = %path.as_ref().display()))
    )]
    pub fn load<P: AsRef<Path>>(path: P) -> Result<Self, crate::Error> {
        let path = path.as_ref().canonicalize()?;
        let config = BiftiPhantom::load(&path)?;
        let base_dir = path.parent().ok_or(crate::Error::NoParent)?;
        Self::load_from_config(config, base_dir)
    }

    #[cfg_attr(feature = "tracing", tracing::instrument(skip_all))]
    pub fn load_from_config<P: AsRef<Path>>(
        config: BiftiPhantom,
        base_dir: P,
    ) -> Result<Self, crate::Error> {
        // A phantom's NIfTIs are shared between tissues (one file, one sub-volume per
        // tissue), and every tissue needs its density twice - as a map and as the
        // resampling weight. Cache the native sub-volumes so each is read exactly once.
        let mut cache = NiftiCache::default();
        let mut tissues = HashMap::new();
        for (name, tissue) in &config.tissues {
            tissues.insert(
                name.clone(),
                Tissue::load(tissue, base_dir.as_ref(), config.reslice_to, &mut cache)?,
            );
        }
        Ok(Self { config, tissues })
    }
}

pub struct Tissue {
    pub density: Volume,
    pub t1: Volume,
    pub t2: Volume,
    pub t2dash: Volume,
    pub adc: Volume,
    pub db0: Volume,
    pub b1_tx: Vec<Volume>,
    pub b1_rx: Vec<Volume>,
}

// ===========================================================================
// Phantom loading internals
// ===========================================================================

impl VolumeData {
    fn from_nifti(slice: nifti::InMemNiftiVolume) -> Result<Self, crate::Error> {
        use nifti::NiftiType as Nt;
        Ok(match slice.data_type() {
            Nt::Uint8 => Self::Uint8(slice.into_nifti_typed_data()?),
            Nt::Uint16 => Self::Uint16(slice.into_nifti_typed_data()?),
            Nt::Uint32 => Self::Uint32(slice.into_nifti_typed_data()?),
            Nt::Uint64 => Self::Uint64(slice.into_nifti_typed_data()?),
            Nt::Int8 => Self::Int8(slice.into_nifti_typed_data()?),
            Nt::Int16 => Self::Int16(slice.into_nifti_typed_data()?),
            Nt::Int32 => Self::Int32(slice.into_nifti_typed_data()?),
            Nt::Int64 => Self::Int64(slice.into_nifti_typed_data()?),
            Nt::Float32 => Self::Float32(slice.into_nifti_typed_data()?),
            Nt::Float64 => Self::Float64(slice.into_nifti_typed_data()?),
            Nt::Complex64 => Self::Complex64(slice.into_nifti_typed_data()?),
            Nt::Complex128 => Self::Complex128(slice.into_nifti_typed_data()?),
            // Not supported: Float128, Complex256, Rgb24, Rgba32
            other => return Err(crate::Error::UnsupportedDataType(format!("{other:?}"))),
        })
    }
}

/// Caches the native (unresampled) NIfTI sub-volumes a phantom refers to, keyed by file and
/// sub-volume index, so a file referenced by several tissues is only read once.
#[derive(Default)]
struct NiftiCache {
    volumes: HashMap<(PathBuf, usize), Rc<Volume>>,
}

impl NiftiCache {
    fn load(&mut self, base_dir: &Path, nifti_ref: &NiftiRef) -> Result<Rc<Volume>, crate::Error> {
        let path = base_dir.join(&nifti_ref.file_name);
        let key = (path.clone(), nifti_ref.tissue_index);
        if let Some(volume) = self.volumes.get(&key) {
            return Ok(Rc::clone(volume));
        }
        let volume = Rc::new(Volume::load_nifti_ref(&path, nifti_ref.tissue_index)?);
        self.volumes.insert(key, Rc::clone(&volume));
        Ok(volume)
    }
}

impl Volume {
    /// Load one sub-volume of a NIfTI file on its **native** grid. Resampling happens later,
    /// in [`Tissue::load`], because it needs the tissue's density map as a weight.
    #[cfg_attr(
        feature = "tracing",
        tracing::instrument(skip_all, fields(file = %path.display(), index))
    )]
    fn load_nifti_ref(path: &Path, index: usize) -> Result<Self, crate::Error> {
        let obj = ReaderStreamedOptions::new().read_file(path)?;

        let header = obj.header();
        let affine = [
            header.srow_x.map(|v| v as f64),
            header.srow_y.map(|v| v as f64),
            header.srow_z.map(|v| v as f64),
        ];

        let mut volume = obj.into_volume();
        let dim = volume.dim();

        if dim.len() != 4 || index >= dim[3] as usize {
            return Err(crate::Error::IndexError {
                index,
                shape: dim.to_vec(),
            });
        }

        let shape = [dim[0] as usize, dim[1] as usize, dim[2] as usize];

        let slice = volume.nth(index).expect("index bounds checked above")?;
        let data = VolumeData::from_nifti(slice)?;

        Ok(Self {
            affine,
            shape,
            data,
        })
    }

    /// Expand a constant (1x1x1) volume onto `reslice_to`'s grid.
    fn expand_to(self, reslice_to: ResliceTo) -> Self {
        let res = reslice_to.resolution;
        Self {
            affine: reslice_to.affine,
            shape: res,
            data: self.data.expand(res[0] * res[1] * res[2]),
        }
    }
}

/// Everything needed to resample one tissue's maps: the shared geometry plus the density
/// weight that keeps intensive properties from being diluted by empty voxels.
struct TissueGrid {
    resampler: Resampler,
    reslice_to: ResliceTo,
    src_affine: [[f64; 4]; 3],
    src_shape: [usize; 3],
    /// `density + eps` on the source grid.
    weight: Vec<f64>,
    /// `sum_j w_j · weight_j` per output voxel - the shared denominator.
    weight_sum: Vec<f64>,
}

impl TissueGrid {
    fn new(density: &Volume, reslice_to: ResliceTo) -> Self {
        let resampler = Resampler::build(density.affine, density.shape, reslice_to);

        // Regularise the weight so a footprint containing no tissue at all still has a
        // well-defined mean - the unweighted mean of its in-bounds taps - instead of 0/0.
        // Scaled by the density's own magnitude, since a `density` map need not be a
        // [0, 1] float map: the spec allows any NIfTI numeric type.
        let values = density.data.to_f64_vec();
        let max = values.iter().copied().fold(0.0f64, f64::max);
        let eps = if max > 0.0 { 1e-6 * max } else { 1.0 };
        let weight: Vec<f64> = values.iter().map(|d| d + eps).collect();

        let weight_sum = resampler.weight_sum(&weight, density.shape);
        Self {
            resampler,
            reslice_to,
            src_affine: density.affine,
            src_shape: density.shape,
            weight,
            weight_sum,
        }
    }

    /// Resample one intensive property, weighted by density.
    ///
    /// The weight only makes sense if the property shares the density map's grid, which
    /// ../../NIFTI.md requires of every NIfTI in a phantom. A non-conforming file falls
    /// back to an unweighted footprint average on its own grid rather than erroring.
    fn resample(&self, native: &Volume) -> Volume {
        let data = if native.affine == self.src_affine && native.shape == self.src_shape {
            map_variant!(&native.data, |v| self.resampler.resample_weighted(
                v,
                self.src_shape,
                &self.weight,
                &self.weight_sum
            ))
        } else {
            warn_grid_mismatch();
            let resampler = Resampler::build(native.affine, native.shape, self.reslice_to);
            map_variant!(&native.data, |v| resampler.resample_plain(v, native.shape))
        };
        Volume {
            affine: self.reslice_to.affine,
            shape: self.reslice_to.resolution,
            data,
        }
    }
}

fn warn_grid_mismatch() {
    #[cfg(feature = "tracing")]
    tracing::warn!(
        "property map does not share the density map's grid (see ../../NIFTI.md); \
         resampling it unweighted, so its values may be diluted near edges"
    );
}

impl Tissue {
    #[cfg_attr(feature = "tracing", tracing::instrument(skip_all))]
    fn load(
        tissue: &BiftiTissue,
        base_dir: &Path,
        reslice_to: Option<ResliceTo>,
        cache: &mut NiftiCache,
    ) -> Result<Self, crate::Error> {
        let density_src = cache.load(base_dir, &tissue.density)?;

        let Some(reslice_to) = reslice_to else {
            // No target grid: every map keeps its own native grid, as before.
            let p = &tissue.properties;
            return Ok(Self {
                density: (*density_src).clone(),
                t1: load_native_property(cache, base_dir, &p.t1)?,
                t2: load_native_property(cache, base_dir, &p.t2)?,
                t2dash: load_native_property(cache, base_dir, &p.t2dash)?,
                adc: load_native_property(cache, base_dir, &p.adc)?,
                db0: load_native_property(cache, base_dir, &p.db0)?,
                b1_tx: load_native_channels(cache, base_dir, &p.b1_tx)?,
                b1_rx: load_native_channels(cache, base_dir, &p.b1_rx)?,
            });
        };

        let grid = TissueGrid::new(&density_src, reslice_to);

        // `density` is a volume fraction - extensive - so it is averaged *unweighted*.
        // Weighting it by itself would renormalise away exactly the partial-volume
        // information the resampling is meant to produce.
        let density = Volume {
            affine: reslice_to.affine,
            shape: reslice_to.resolution,
            data: map_variant!(&density_src.data, |v| grid
                .resampler
                .resample_plain(v, density_src.shape)),
        };

        let p = &tissue.properties;
        Ok(Self {
            density,
            t1: load_resampled_property(cache, base_dir, &p.t1, &grid)?,
            t2: load_resampled_property(cache, base_dir, &p.t2, &grid)?,
            t2dash: load_resampled_property(cache, base_dir, &p.t2dash, &grid)?,
            adc: load_resampled_property(cache, base_dir, &p.adc, &grid)?,
            db0: load_resampled_property(cache, base_dir, &p.db0, &grid)?,
            b1_tx: load_resampled_channels(cache, base_dir, &p.b1_tx, &grid)?,
            b1_rx: load_resampled_channels(cache, base_dir, &p.b1_rx, &grid)?,
        })
    }
}

/// Resolve a property on its native grid (the `reslice_to`-absent path).
fn load_native_property(
    cache: &mut NiftiCache,
    base_dir: &Path,
    property: &TissueProperty,
) -> Result<Volume, crate::Error> {
    Ok(match property {
        TissueProperty::Value(value) => Volume::single_voxel(*value),
        TissueProperty::Ref(nifti_ref) => (*cache.load(base_dir, nifti_ref)?).clone(),
        TissueProperty::Mapping(mapping) => {
            let volume = (*cache.load(base_dir, &mapping.file)?).clone();
            crate::eval::eval_mapping_func(volume, &mapping.func)?
        }
    })
}

/// Resolve every channel of a `B1+`/`B1-` array on its native grid.
fn load_native_channels(
    cache: &mut NiftiCache,
    base_dir: &Path,
    channels: &[TissueProperty],
) -> Result<Vec<Volume>, crate::Error> {
    channels
        .iter()
        .map(|ch| load_native_property(cache, base_dir, ch))
        .collect()
}

/// Resolve every channel of a `B1+`/`B1-` array onto the tissue's target grid.
fn load_resampled_channels(
    cache: &mut NiftiCache,
    base_dir: &Path,
    channels: &[TissueProperty],
    grid: &TissueGrid,
) -> Result<Vec<Volume>, crate::Error> {
    channels
        .iter()
        .map(|ch| load_resampled_property(cache, base_dir, ch, grid))
        .collect()
}

/// Resolve a property and resample it onto the tissue's target grid, weighted by density.
fn load_resampled_property(
    cache: &mut NiftiCache,
    base_dir: &Path,
    property: &TissueProperty,
    grid: &TissueGrid,
) -> Result<Volume, crate::Error> {
    Ok(match property {
        // A uniform value needs no resampling - just fill the target grid.
        TissueProperty::Value(value) => Volume::single_voxel(*value).expand_to(grid.reslice_to),
        TissueProperty::Ref(nifti_ref) => {
            let native = cache.load(base_dir, nifti_ref)?;
            grid.resample(&native)
        }
        TissueProperty::Mapping(mapping) => {
            let native = cache.load(base_dir, &mapping.file)?;
            let resampled = grid.resample(&native);
            crate::eval::eval_mapping_func(resampled, &mapping.func)?
        }
    })
}

// ===========================================================================
// Affine helpers
// ===========================================================================

/// `a ∘ b`: the affine applying `b` first, then `a`.
pub(crate) fn compose_affine(a: [[f64; 4]; 3], b: [[f64; 4]; 3]) -> [[f64; 4]; 3] {
    let mut out = [[0.0; 4]; 3];
    for (i, row) in out.iter_mut().enumerate() {
        for (j, cell) in row.iter_mut().enumerate().take(3) {
            *cell = a[i][0] * b[0][j] + a[i][1] * b[1][j] + a[i][2] * b[2][j];
        }
        row[3] = a[i][0] * b[0][3] + a[i][1] * b[1][3] + a[i][2] * b[2][3] + a[i][3];
    }
    out
}

pub(crate) fn invert_affine(a: [[f64; 4]; 3]) -> [[f64; 4]; 3] {
    let inv_det = 1.0
        / (a[0][0] * (a[1][1] * a[2][2] - a[1][2] * a[2][1])
            - a[0][1] * (a[1][0] * a[2][2] - a[1][2] * a[2][0])
            + a[0][2] * (a[1][0] * a[2][1] - a[1][1] * a[2][0]));

    let r = [
        [
            (a[1][1] * a[2][2] - a[1][2] * a[2][1]) * inv_det,
            (a[0][2] * a[2][1] - a[0][1] * a[2][2]) * inv_det,
            (a[0][1] * a[1][2] - a[0][2] * a[1][1]) * inv_det,
        ],
        [
            (a[1][2] * a[2][0] - a[1][0] * a[2][2]) * inv_det,
            (a[0][0] * a[2][2] - a[0][2] * a[2][0]) * inv_det,
            (a[0][2] * a[1][0] - a[0][0] * a[1][2]) * inv_det,
        ],
        [
            (a[1][0] * a[2][1] - a[1][1] * a[2][0]) * inv_det,
            (a[0][1] * a[2][0] - a[0][0] * a[2][1]) * inv_det,
            (a[0][0] * a[1][1] - a[0][1] * a[1][0]) * inv_det,
        ],
    ];

    // inverse 3x4 matrix of the input - offset is mapped to the new system
    [
        [
            r[0][0],
            r[0][1],
            r[0][2],
            -(r[0][0] * a[0][3] + r[0][1] * a[1][3] + r[0][2] * a[2][3]),
        ],
        [
            r[1][0],
            r[1][1],
            r[1][2],
            -(r[1][0] * a[0][3] + r[1][1] * a[1][3] + r[1][2] * a[2][3]),
        ],
        [
            r[2][0],
            r[2][1],
            r[2][2],
            -(r[2][0] * a[0][3] + r[2][1] * a[1][3] + r[2][2] * a[2][3]),
        ],
    ]
}

File: python/bifti/src/bifti/loader.py
# Reference loader for NIfTI phantoms: turns a phantom (parsed by nifti_phantom)
# into plain NumPy arrays. A readable example for porting to your own library -
# not optimised or feature-complete. Deps: numpy, nibabel (and torch, used for
# resampling when it is installed). See ../SPEC.md for the format and DEMO.md
# for usage.

from __future__ import annotations

from dataclasses import dataclass
from functools import lru_cache
from pathlib import Path

import numpy as np
import nibabel

from .resample import Resampler, density_weight, warn_grid_mismatch
from .phantom import (
    BiftiPhantom,
    BiftiTissue,
    NiftiRef,
    NiftiMapping,
    ResliceTo,
)


# ===========================================================================
# Public entry points
# ===========================================================================


@dataclass
class NumpyPhantom:
    """A loaded bifti phantom + maps, represented as dict of [`NumpyTissue`]"""

    config: BiftiPhantom
    tissues: dict[str, NumpyTissue]

    @classmethod
    def load(cls, path: Path | str):
        """Load a complete phantom from its ``phantom.json``.

        NIfTI files are resolved relative to the JSON file's directory, matching the
        folder convention in ``../SPEC.md``.
        """
        path = Path(path)
        config = BiftiPhantom.load(Path(path))
        return cls.load_from_config(config, base_dir=path.parent)

    @classmethod
    def load_from_config(cls, config: BiftiPhantom, base_dir: Path | str):
        """Load all tissues of an already-parsed config from ``base_dir``.

        Useful if you want to tweak the config in memory before loading the data.
        """
        return cls(
            config=config,
            tissues={
                name: load_tissue(tissue, Path(base_dir), config.reslice_to)
                for name, tissue in config.tissues.items()
            },
        )


@dataclass
class NumpyTissue:
    """A single tissue with every property resolved to a NumPy array.

    All scalar properties are expanded to full arrays so that downstream code
    can treat uniform and spatially-varying tissues the same way. Every array
    shares the same 3D ``shape`` and the same voxel-to-world ``affine``.

    Units follow the spec (see ``../JSON.md``): T1/T2/T2' in seconds, ADC in
    1e-3 mm^2/s, dB0 in Hz, B1+/B1- relative, density a volume fraction.
    """

    density: np.ndarray  # (X, Y, Z)            volume fraction
    T1: np.ndarray  #      (X, Y, Z)            seconds
    T2: np.ndarray  #      (X, Y, Z)            seconds
    T2dash: np.ndarray  #  (X, Y, Z)            seconds
    ADC: np.ndarray  #     (X, Y, Z)            1e-3 mm^2/s
    dB0: np.ndarray  #     (X, Y, Z)            Hz
    B1_tx: np.ndarray  #   (channels, X, Y, Z)  relative transmit field
    B1_rx: np.ndarray  #   (channels, X, Y, Z)  relative receive field
    resliced: ResliceTo  # Affine + resolution this phantom is resliced to

    @property
    def shape(self) -> list[int]:
        return self.resliced.resolution

    @property
    def affine(self) -> list[list[float]]:
        return self.resliced.affine


# ===========================================================================
# Phantom loading internals
# ===========================================================================


def load_tissue(
    tissue: BiftiTissue,
    base_dir: Path | str,
    reslice_to: ResliceTo | None = None,
) -> NumpyTissue:
    """Load one tissue, resolving every property to a NumPy array.

    The output grid is ``reslice_to`` if given, otherwise the density map's own grid - so
    every other map is brought onto the density resolution and affine (see ``../JSON.md``
    -> ``reslice_to``). The spec requires a phantom's NIfTIs to already share that grid, so
    for conforming data the implicit resampling is a no-op.

    Resampling averages each output voxel over the source region it covers, weighted by
    this tissue's ``density`` - see :mod:`bifti.resample` for why the weighting matters.
    """
    base_dir = Path(base_dir)

    # The density map defines the source grid *and* supplies the resampling weight, so it
    # is loaded natively first and resampled separately from everything else.
    density_src, src_affine = load_file_ref_noreslice(base_dir, tissue.density)
    src_shape = tuple(density_src.shape)

    if reslice_to is None:
        reslice_to = ResliceTo(affine=src_affine, resolution=list(src_shape))

    resampler = Resampler.build(
        src_affine, src_shape, reslice_to.affine, reslice_to.resolution
    )

    # `density` is a volume fraction - extensive - so it is averaged *unweighted*.
    # Weighting it by itself would renormalise away exactly the partial-volume information
    # the resampling is meant to produce.
    density = resampler.resample_plain(density_src)

    # The denominator every intensive property of this tissue shares.
    weight = density_weight(density_src)
    weight_sum = resampler.weight_sum(weight)

    def prop(cfg, name) -> np.ndarray:
        return load_property(
            cfg, base_dir, reslice_to, resampler, weight, weight_sum, src_affine,
            src_shape, name,
        )

    def channels(cfgs, name) -> np.ndarray:
        return np.stack(
            [prop(ch, f"{name}[{i}]") for i, ch in enumerate(cfgs)], axis=0
        )

    return NumpyTissue(
        density=density,
        T1=prop(tissue.T1, "T1"),
        T2=prop(tissue.T2, "T2"),
        T2dash=prop(tissue.T2dash, "T2'"),
        ADC=prop(tissue.ADC, "ADC"),
        dB0=prop(tissue.dB0, "dB0"),
        B1_tx=channels(tissue.B1_tx, "B1+"),
        B1_rx=channels(tissue.B1_rx, "B1-"),
        resliced=reslice_to,
    )


def load_property(
    config: float | NiftiRef | NiftiMapping,
    base_dir: Path,
    reslice_to: ResliceTo,
    resampler: Resampler,
    weight: np.ndarray,
    weight_sum: np.ndarray,
    src_affine: list[list[float]],
    src_shape: tuple[int, ...],
    name: str,
) -> np.ndarray:
    """Resolve one "scalar-or-map" property to a 3D array (../JSON.md).

    1. a number      -> a uniform array of ``shape`` filled with that value;
    2. a NIfTI ref   -> the referenced sub-volume, density-weighted onto the target grid;
    3. a transformed -> case 2 with ``func`` applied per voxel.

    Maps come back already on the output grid, so a ``func`` and its ``x_*`` statistics act
    on the resampled values.
    """
    if isinstance(config, (int, float)):
        # A uniform value needs no resampling - just fill the target grid.
        return np.full(reslice_to.resolution, float(config), dtype=np.float64)
    if isinstance(config, NiftiRef):
        ref = config
    elif isinstance(config, NiftiMapping):
        ref = config.file
    else:
        raise TypeError(
            f"property must be a number, NiftiRef or NiftiMapping, got {type(config)}"
        )

    native, native_affine = load_file_ref_noreslice(base_dir, ref)
    if tuple(native.shape) == src_shape and np.allclose(native_affine, src_affine):
        data = resampler.resample_weighted(native, weight, weight_sum)
    else:
        # ../NIFTI.md requires every NIfTI of a phantom to share a grid. A non-conforming
        # file cannot use the density weight, so fall back to an unweighted average.
        warn_grid_mismatch(name)
        other = Resampler.build(
            native_affine, native.shape, reslice_to.affine, reslice_to.resolution
        )
        data = other.resample_plain(native)

    if isinstance(config, NiftiMapping):
        return eval_expr(config.func, data)
    return data


# ===========================================================================
# NIfTI file access
# ===========================================================================


def load_file_ref_noreslice(
    base_dir: Path, ref: NiftiRef
) -> tuple[np.ndarray, list[list[float]]]:
    """Load the sub-volume named by ``ref`` on its native grid.

    ``ref`` is a ``"<file>[<index>]"`` reference; ``index`` selects along the
    NIfTI's 4th (tissue) dimension. Returns the 3D sub-volume and the file's own
    3x4 affine (the upper rows of its 4x4 voxel-to-world transform).
    """
    data, affine = _load_nifti(base_dir, ref.file_name)
    return data[:, :, :, ref.tissue_index], affine


# Avoid re-loading NIfTIs for every tissue by caching the native (unresampled) file.
# Resampling can no longer be cached per file: it depends on the *tissue's* density map,
# which generally lives in a different file. Caching the raw load instead is if anything a
# stronger cache, since it no longer varies with the target grid.
@lru_cache(maxsize=20)
def _load_nifti(
    base_dir: Path, file_name: Path
) -> tuple[np.ndarray, list[list[float]]]:
    # A reference may be relative to the phantom directory or an absolute path.
    path = file_name if file_name.is_absolute() else (base_dir / file_name).resolve()
    img = nibabel.load(path)
    assert isinstance(img, nibabel.Nifti1Image)
    assert len(img.shape) == 4

    data = np.asarray(img.dataobj, dtype=np.float64)
    sform = img.get_sform()  # full 4x4 voxel-to-world (RAS+, mm); see ../NIFTI.md
    return data, sform[:3].tolist()


# ===========================================================================
# `func` transforms (../JSON.md -> "Transformed reference")
# ===========================================================================


def eval_expr(func: str, data: np.ndarray) -> np.ndarray:
    """Apply a ``func`` transform to a voxel array.

    ``x`` is the per-voxel value; ``x_min``/``x_max``/``x_mean``/``x_std`` are
    scalar statistics of the whole volume.

    NOTE: ``func`` comes straight from the phantom file and is run with ``eval``
    here for brevity, so only load phantoms you trust. The spec restricts it to
    numbers, ``+ - * /``, parentheses and the ``x*`` variables - a hardened
    implementation should parse exactly that grammar instead.
    """
    from warnings import warn

    warn(f"Executing mapping function: '{func}' (possible RCE!)")
    return eval(
        func,
        {"__builtins__": None},
        {
            "x": data,
            "x_min": data.min(),
            "x_max": data.max(),
            "x_mean": data.mean(),
            "x_std": data.std(),
        },
    )


# ===========================================================================
# Example usage
# ===========================================================================

if __name__ == "__main__":
    import sys

    if len(sys.argv) != 2:
        print("usage: python nifti_loader.py <path/to/phantom.json>")
        raise SystemExit(2)

    phantom = NumpyPhantom.load(sys.argv[1])
    for name, tissue in phantom.tissues.items():
        print(f"{name}: shape={tissue.shape}, B1+ channels={tissue.B1_tx.shape[0]}")
        print(
            f"    T1 mean={np.nanmean(tissue.T1):.4g} s, dB0 range="
            f"[{tissue.dB0.min():.4g}, {tissue.dB0.max():.4g}] Hz"
        )
